### DISCUSSION TITLE
FYI playing around with pint and pydantic --discussion

### DISCUSSION THREAD

**Discussion by uellue** [hyperspy/hyperspy#3431] [Ideas]
In LiberTEM we have been brainstorming on how metadata can be specified and validated in such a way that it works for both software developers and users. We found it pretty challenging to make metadata consistent even within our own project.

We've [combined `pydantic` and `pint`](https://github.com/LiberTEM/LiberTEM-schema/blob/main/src/libertem_schema/__init__.py) so that a schema can be defined easily:

```python

class Simple4DSTEMParams(BaseModel):
    '''
    Basic calibration parameters of a strongly simplified model
    of a 4D STEM experiment.

    See https://github.com/LiberTEM/Microscope-Calibration
    and https://arxiv.org/abs/2403.08538
    for the technical details.
    '''
    overfocus: Length
    scan_pixel_pitch: Length
    camera_length: Length
    detector_pixel_pitch: Length
    semiconv: Angle
    cy: Pixel
    cx: Pixel
    scan_rotation: Angle
    flip_y: bool
```

...and also instantiated easily:

```python
params = Simple4DSTEMParams(
    overfocus=0.0015 * ureg.meter,
    scan_pixel_pitch=10000 * ureg.angstrom,
    camera_length=0.15 * ureg.meter,
    detector_pixel_pitch=0.000050 * ureg.meter,
    semiconv=0.020 * ureg.radian,
    scan_rotation=330. * ureg.degree,
    flip_y=False,
    cy=(32 - 2) * ureg.pixel,
    cx=(32 - 2) * ureg.pixel,
)
```

Missing parameters or dimensionality mismatch will lead to a validation error when trying to instantiate. See [the tests](https://github.com/LiberTEM/LiberTEM-schema/blob/c096d5337f21c78232134ad9d9af19b8405b1992/tests/test_schemas.py#L1) for more usage and error scenarios. The instance can be serialized as JSON, the schema as JSON Schema, and then both can be also validated with, for example, `jsonschema`. That means it is portable between programming languages and anything that maps from and to JSON.

For the time being we are just trying to make parameters consistent within LiberTEM (next step is starting to actually use it there), but in the future one could think about translating equivalent parameters, or perhaps define the HyperSpy metadata schema like that? I was looking if I could find some schema definition from HyperSpy so that we can re-use it, but it seems to be mostly defined by convention? Please correct me if I am wrong!

Long story short, since this is, in the end, about interoperability, I figured it could be nice to share it and start a discussion!

CC @sk1p @magnunor @matbryan52

**Comment by CSSFrancis** [comment]
@uellue This is a great discussion.  There has been some previous discussion on [metadata](https://github.com/hyperspy/hyperspy/issues/2095) but we have previously run into a bit of a roadblock defining how metadata should be handled.  I think it's important to have a standard but not be too ridged such that you continue give people flexibility.  

There is added complexity when you are trying to be domain agnostic.  The example that always comes to mind of this is how the xray community defines reciporical space in terms of "q" where the electron diffraction community uses "k" both have units of angstroms^-1 but they are off by a factor of 2pi!  Hyperspy does use pint and you can do something like `s.isig["1nm":"2nm"]` to slice using real units.  Although `s.isig[1 *nm : 2 *nm ]` might be more pythonic. 

I've thought of this a bit downstream in pyxem and made a [Calibration](https://github.com/pyxem/pyxem/blob/main/pyxem/utils/calibration.py) class.  Somethings that might be worth thinking of:

- What if your pivot point correction is wrong. Should that be corrected (currently what we do in pyxem) or can an array of centers be passed.
- What happens when you change the scale?  Does the center value also update? 
- Does your reciporical space scale depend on your camera length? It tecnically should be non linear when expressed in inverse nm/angstroms and should be linear when expressed in terms of mrad. That is seldom done in electron microscopy but essential for xray diffraction. 
- How do you express non-linear scales? 
- Could you extend this to something that has arbitrary scan positions? 
- Do you need to make a new class for 5D STEM metadata? What about 6D or ND?

A couple of other thoughts:

1. How hard is it to remove the LiberTEM dependancy? 
2. Can you reliscense MIT or some other permissive liscense.  Direct Electron (and other microscope vendors) would love an open source metadata format to write to.

**Reply by sk1p** [reply]
> I've thought of this a bit downstream in pyxem and made a [Calibration](https://github.com/pyxem/pyxem/blob/main/pyxem/utils/calibration.py) class

Thanks for the link. I noticed that for metadata handling, there are `signal.metadata.get_item(...)` calls sprinkled in the code - is there an overview (per signal-class?) somewhere? The closest I can find are the docs of [LumiSpy](https://docs.lumispy.org/en/latest/user_guide/metadata_structure.html) and [eXSpy](https://hyperspy.org/exspy/user_guide/metadata_structure.html) - is there such a page for pyxem, too? Is the definition of, for example, `Acquisition_instrument.TEM`, shared between eXSpy and  pyxem?


For some added context, while @uellue is now mostly working on defining metadata as input parameters for processing methods, my own interest is that LiberTEM-live will soon gain a zarr writer, which _could_ emit `zspy` data sets with metadata that hyperspy/pyxem/... directly understand, instead of rolling our own format. For the data itself, this should be easy enough, from the description in `rosettasciio`, but metadata is another story... let me know if this should be a separate discussion btw.

>  I think it's important to have a standard but not be too ridged such that you continue give people flexibility.

Completely agree - for me, it's mostly about having a mapping for common (core?) parameters/calibrations. If I have a `wavelength` available, where (and in what type+unit) would I write this so `pyxem`/`hyperspy`/... can pick it up? (and should I write `wavelength` vs. `beam_energy` vs `accelerating_voltage`?)

If the user wants to put in additional metadata, we won't stop them. Just the opposite: we will provide methods to attach additional custom metadata to the `zarr` array/group (before and after acquiring the data), which could be things like in-situ parameters, vendor-specific metadata, or anything they desire.

One point from your list:

> * What if your pivot point correction is wrong. Should that be corrected (currently what we do in pyxem) or can an array of centers be passed.

We go into the direction of the latter (cc @sivborg). In case of metadata in a file format, I think it should be possible to describe the data as-recorded. If there is an array of centers attached to the data (for example by running a calibration routine before acquisition), pyxem could use that as input to perform its pivot point correction.

**Reply by CSSFrancis** [reply]
> Thanks for the link. I noticed that for metadata handling, there are `signal.metadata.get_item(...)` calls sprinkled in the code - is there an overview (per signal-class?) somewhere? The closest I can find are the docs of [LumiSpy](https://docs.lumispy.org/en/latest/user_guide/metadata_structure.html) and [eXSpy](https://hyperspy.org/exspy/user_guide/metadata_structure.html) - is there such a page for pyxem, too? Is the definition of, for example, `Acquisition_instrument.TEM`, shared between eXSpy and pyxem?

For the most part `pyxem` has been quite bad at defining any such strucuture for a couple of reasons. We try our best to follow the metadata strucuture (kind of) [suggested](https://hyperspy.org/hyperspy-doc/current/reference/metadata.html#metadata-structure) by `hyperspy`. My personal opinion is that currently it's a bit clunky to work with the metadata tree and I would rather [set ](https://github.com/hyperspy/hyperspy/pull/3123) the beam energy using something like `s.calibration.beam_energy= 200`.  Similarly if someone sets the beam energy, then the wavelength should also be [set](https://github.com/pyxem/pyxem/blob/47721292a5961d54e7f054545fa4e614dfc7c874/pyxem/utils/calibration.py#L139).

As previously stated, should the `camera_length` be set by the TEM or should it be connected to the scale?  I've just found that these aren't super easy relationships to describe simply and people have different calibration workflows. 



The ones that are a bit hidden that need to be more explicit from Hyperspy's end:

- s.metadata._HyperSpy.navigator (accessible via the `.navigator` attribute)
- s.metadata.Signal.quantity ( Sets the units for the color bar)
- s.metadata.General.title (Sets the title for a plot)

> For some added context, while @uellue is now mostly working on defining metadata as input parameters for processing methods, my own interest is that LiberTEM-live will soon gain a zarr writer, which _could_ emit `zspy` data sets with metadata that hyperspy/pyxem/... directly understand, instead of rolling our own format. For the data itself, this should be easy enough, from the description in `rosettasciio`, but metadata is another story... let me know if this should be a separate discussion btw.

I can define a metadata schema for pyxem based on what we currently use and add that to the documentation.  I'll ping you when I make the PR. This is also something that Direct Electron is working on so it would be good to get everyone on the same page.

> > I think it's important to have a standard but not be too ridged such that you continue give people flexibility.
> 
> Completely agree - for me, it's mostly about having a mapping for common (core?) parameters/calibrations. If I have a `wavelength` available, where (and in what type+unit) would I write this so `pyxem`/`hyperspy`/... can pick it up? (and should I write `wavelength` vs. `beam_energy` vs `accelerating_voltage`?)

I like that :)  For reference currently:

`"When a leaf contains a quantity that is not dimensionless, the units can be given in an extra leaf with the same label followed by the “_units” suffix. For example, an “energy” leaf should be accompanied by an “energy_units” leaf""` Although this proceedure isn't really followed in pyxem currently... 

> 
> If the user wants to put in additional metadata, we won't stop them. Just the opposite: we will provide methods to attach additional custom metadata to the `zarr` array/group (before and after acquiring the data), which could be things like in-situ parameters, vendor-specific metadata, or anything they desire.

Perfect :)

> One point from your list:
> 
> > * What if your pivot point correction is wrong. Should that be corrected (currently what we do in pyxem) or can an array of centers be passed.
> 
> We go into the direction of the latter (cc @sivborg). In case of metadata in a file format, I think it should be possible to describe the data as-recorded. If there is an array of centers attached to the data (for example by running a calibration routine before acquisition), pyxem could use that as input to perform its pivot point correction.

Perfect.  We should probably have that as an additional metadata attribute that gets stored after the `get_direct_beam_postion` function.

**Reply by uellue** [reply]
Lots to reply, so I'll copy some pieces here!

> I think it's important to have a standard but not be too ridged such that you continue give people flexibility.

Absolutely agree! That's why I've tried to make defining schemas very easy and clear so that people can create and modify them as needed. The current one grew from "we need to pass some parameters" to sth we can (I think) use throughout the project, and I imagine that other schemas will grow like that as well.

> What if your pivot point correction is wrong. (...)

That's a very good point, metadata that can capture the experiment properly and still translate to all kinds of analyses! In my opinion one needs a "digital twin" of the microscope that can capture it in all detail, and then fit simplified models to it. In https://github.com/LiberTEM/Microscope-Calibration we have a very simple prototype of that which translates between a simple human-compatible model, an instrument simulation with simulated scan and descan coils, and a math kernel. That's a bit of a longer story, the repo links to an arXiv preprint with the writeup. :-)

I wanted to show not so much the exact parameters of `Simple4DSTEMParams`, but using pydantic and pint to very easily define a generic schema that can be validated automatically, is easy to use, not tied to any project specifics, and allows cross-platform interoperability.

> I can define a metadata schema for pyxem based on what we currently use and add that to the documentation.

That would be awesome! If you are interested, feel free to re-use the pydantic+pint approach (MIT license for a reason!). I'm considering setting this up as a separate project since people in the pint issue tracker found it interesting https://github.com/hgrecco/pint/issues/1166#issuecomment-2326870178, just having a lot on my TODO list at the moment... If you'd like to use the same pint+pydantic machinery we'd have a good case to bump it towards the head of the TODO queue. Maybe we can also see if/how we can make the schema itself interoperable between pyxem and LiberTEM? Re-using pyxem's/Hyperspy's metadata schema was my plan A, I just didn't find it in an accessible form.

>  The example that always comes to mind of this is how the xray community defines reciporical space in terms of "q" where the electron diffraction community uses "k" both have units of angstroms^-1 but they are off by a factor of 2pi!

Yes, these ones are tricky! IMO one can address this with descriptive text, examples and test routines. First step would be not calling it `q` or `k`, but sth a bit more descriptive! Probably we can connect with the "ontology" efforts here.

**Comment by sk1p** [comment]
Just to quickly address these two points:

> 1. How hard is it to remove the LiberTEM dependancy?

For the schema repo, there is none: https://github.com/LiberTEM/LiberTEM-schema/blob/c096d5337f21c78232134ad9d9af19b8405b1992/pyproject.toml#L12-L16

> 2. Can you reliscense MIT or some other permissive liscense.  Direct Electron (and other microscope vendors) would love an open source metadata format to write to.

Already is: https://github.com/LiberTEM/LiberTEM-schema/blob/main/LICENSE :)

**Comment by CSSFrancis** [comment]
@francisco-dlp @jat255 @bbammes @jlaehne
